### PR TITLE
[AArch64] Optimize when storing symmetry constants

### PR DESCRIPTION
**Pull request by ParkHanbum**
This change looks for instructions of storing symmetric constants
instruction 32-bit units. usually consisting of several 'MOV' and
one or less 'ORR'.

If found, load only the lower 32-bit constant and change it to copy
and save to the upper 32-bit using the 'STP' instruction.

For example:
  renamable $x8 = MOVZXi 49370, 0
  renamable $x8 = MOVKXi $x8, 320, 16
  renamable $x8 = ORRXrs $x8, $x8, 32
  STRXui killed renamable $x8, killed renamable $x0, 0
becomes
  $w8 = MOVZWi 49370, 0
  $w8 = MOVKWi $w8, 320, 16
  STPWi killed renamable $w8, killed renamable $w8, killed renamable $x0, 0

related issue : https://github.com/llvm/llvm-project/issues/51483

### CODE REVIEW

**Comment by ParkHanbum** [comment]
@davemgreen would you review this, please?

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 1]
```diff
@@ -2252,6 +2259,167 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+      LiveRegUnits::accumulateUsedDefed(MI, ModifiedRegUnits, UsedRegUnits,
+                                        TRI);
+      if (!ModifiedRegUnits.available(BaseReg) ||
+          !UsedRegUnits.available(BaseReg))
+        break;
+      continue;
+    }
+
+    unsigned Opc = MI.getOpcode();
+    if (Opc == AArch64::ORRXrs) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2384]: One advantage to matching from the ORRXrs is that we don't actually need to know the exact instruction feeding it, so long as we know the top 32bits are 0 then we should be able to do the transform. Would it be possible to capture the cases you are interested in by detecting the MOVZXi/MOVKXi as having zero upper bits? We should be able to add a few extra instructions that we know don't set top bits too.

Comment (ParkHanbum) [reply llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2384]: as you know, I wrote the code with the case of not including `ORR` in mind. 

That is, I wanted to be able to optimize for the case where arbitrary code loads symmetric 64-bit values into registers, even if they are not contiguous 64-bit values. 

But I also realized that I was overextending into gray areas that I didn't know about. as you said, I would change to a pattern targeted at `ORR`.

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 2]
```diff
@@ -2252,6 +2260,159 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+
+  Register DstRegW =
+      TRI->getSubReg(FirstMovI->getOperand(0).getReg(), AArch64::sub_32);
+  int Lower = Accumulated & Mask;
+  if (Lower) {
+    BuildMI(*MBB, FirstMovI, FirstMovI->getDebugLoc(),
+            TII->get(AArch64::MOVZWi), DstRegW)
+        .addImm(Lower)
+        .addImm(0);
+    Lower = Accumulated >> 16 & Mask;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2320]: (Accumulated >> 16) & Mask;

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 3]
```diff
@@ -2252,6 +2260,159 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+    int SuccIndex, int Accumulated) {
+  MachineBasicBlock::iterator I = MI.getIterator();
+  MachineBasicBlock::iterator E = I->getParent()->end();
+  MachineBasicBlock::iterator NextI = next_nodbg(I, E);
+  MachineBasicBlock::iterator FirstMovI;
+  MachineBasicBlock *MBB = MI.getParent();
+  uint64_t Mask = 0xFFFFUL;
+  int Index = 0;
+
+  for (auto MI = MIs.begin(), E = MIs.end(); MI != E; ++MI, Index++) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2304]: If the input is already 2 Movs+or then it could just remove the or and replace it with a subreg extract.

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 4]
```diff
@@ -2252,6 +2260,159 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+      MIs.push_back(MBBI);
+      continue;
+    }
+    unsigned ValueOrder = Opc == AArch64::MOVZXi ? 1 : 2;
+    MachineOperand Value = MI.getOperand(ValueOrder);
+    MachineOperand Shift = MI.getOperand(ValueOrder + 1);
+    if (!Value.isImm() || !Shift.isImm())
+      return false;
+
+    IValue = Value.getImm();
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2397]: `uint64_t IValue = ..`
`uint64_t IShift = ..`

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 5]
```diff
@@ -2252,6 +2260,159 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+    IShift = Shift.getImm();
+    Accumulated -= (Accumulated & (Mask << IShift));
+    Accumulated += (IValue << IShift);
+    MIs.push_back(MBBI);
+    if ((Accumulated != 0) &&
+        (((Accumulated >> 32) == (Accumulated & 0xffffffffULL)) ||
+         ((hasORR && Accumulated >> 32 == 0))))
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2404]: Extra brackets can be removed.

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 6]
```diff
@@ -2252,6 +2260,159 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+      SuccIndex = MIs.size();
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2405]: Should this break once it has found a valid immediate?

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp [hunk 7]
```diff
@@ -2252,6 +2260,159 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
...
+    MBBI = prev_nodbg(MBBI, B);
+    MachineInstr &MI = *MBBI;
+    if (!MI.isTransient())
+      ++Count;
+    if (!isSymmetric(MI, BaseReg)) {
+      LiveRegUnits::accumulateUsedDefed(MI, ModifiedRegUnits, UsedRegUnits,
+                                        TRI);
+      if (!ModifiedRegUnits.available(BaseReg) ||
+          !UsedRegUnits.available(BaseReg))
+        break;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2381]: Should this return false? So when we break out of the loop we know we are in a valid situation.

**Review by davemgreen — COMMENTED** [review]
Do you have access to an AArch64 machine that you could run the llvm-test-suite on to check the correctness? Or perhaps a bootstrap?

**Comment by ParkHanbum** [comment]
@davemgreen  I tested in M1 macbook by using ninja check-all. let me know if there was better way to test this to.

**Comment by davemgreen** [comment]
The lit tests are mostly unit level tests, and it can be good to run end-to-end tests too in order to increase the test coverage. There is some details in https://www.llvm.org/docs/TestSuiteGuide.html

**Comment by ParkHanbum** [comment]
thanks for notice!

File: llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp (hunk 3 above)
Comment (ParkHanbum) [reply llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:2304]: can I replace operand register of instruction without change its opcode?

**Comment by ParkHanbum** [comment]
this is result of running test-suite

Failed Tests (1):
  test-suite :: SingleSource/Regression/C/gcc-c-torture/execute/GCC-C-execute-alloca-1.test


Testing Time: 2607.14s

Total Discovered Tests: 3443
  Passed: 3442 (99.97%)
  Failed:    1 (0.03%)

**Comment by davemgreen** [comment]
Hi - I was taking a look at what might be wrong. I think the script was compiling with `-mcpu=native -Ofast -flto`, and LTO can be difficult to debug. Luckily it looks like the assembly didn't change too much though, and the difference might be down to the offsets of the ldp's needing to be updated. Can you add some tests with non-zero offsets and make sure they look OK?
```
define void @test_store_0x1234567812345678_o(ptr %x) {
; CHECK-LABEL: test_store_0x1234567812345678_o:       
; CHECK:       // %bb.0:                              
; CHECK-NEXT:    mov x8, #22136 // =0x5678            
; CHECK-NEXT:    movk x8, #4660, lsl #16              
; CHECK-NEXT:    stp w8, w8, [x0, #32]                
; CHECK-NEXT:    ret                                  
  %g = getelementptr i64, ptr %x, i64 4               
  store i64 u0x1234567812345678, ptr %g               
  ret void                                            
}                                                     
```

**Comment by ParkHanbum** [comment]
I realized that what problem in my code. STP operand is not properly. I'll fix it ASAP, but I don't understand why my test is passed. something wrong in my developement enviroment.

**Comment by ParkHanbum** [comment]
I got the pass in test-suite again. 
in changed environment from MacOSX to Linux with docker. 

```
/work/llvm/build-linux/test-suite-build/../../test-suite/utils/compare.py:206: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  name0 = names[0]
Program                                       exec_time 
                                              results   
LoopVector...meChecks4PointersDBeforeA/1000   1747554.95
LoopVector...imeChecks4PointersDAfterA/1000    394555.76
LoopVector...meChecks4PointersDEqualsA/1000    391361.31
LoopVector...tersAllDisjointIncreasing/1000    389818.10
LoopVector...tersAllDisjointDecreasing/1000    387404.53
LCALS/Subs...test:BM_MAT_X_MAT_LAMBDA/44217    200970.88
LCALS/Subs...aw.test:BM_MAT_X_MAT_RAW/44217    185668.65
harris/har...est:BENCHMARK_HARRIS/2048/2048     72338.17
ImageProce...t:BENCHMARK_GAUSSIAN_BLUR/1024     37052.50
LoopVector...timeChecks4PointersDBeforeA/32     33376.06
LoopVector...LoopFrom_uint64_t_To_uint32_t_     22920.46
LoopVector...VW16From_uint64_t_To_uint32_t_     20571.81
LoopVector...hVW8From_uint64_t_To_uint32_t_     19181.25
LoopVector...hVW16From_uint64_t_To_uint8_t_     17468.18
LoopVector...LoopFrom_uint64_t_To_uint16_t_     17304.87
          exec_time
run         results
count  4.260000e+03
mean   1.104730e+03
std    2.967942e+04
min    0.000000e+00
25%    4.000000e-04
50%    1.360000e-02
75%    9.477289e-02
max    1.747555e+06

root@6edef4f7cbda:/work/llvm/build-linux/test-suite-build# uname -a
Linux 6edef4f7cbda 6.4.16-linuxkit #1 SMP PREEMPT Thu Nov 16 10:49:20 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
```

@davemgreen please let me know still problem occur.

File: llvm/test/CodeGen/AArch64/movimm-expand-ldst.ll [hunk 8]
```diff
@@ -93,3 +93,134 @@ define i64 @testuu0xf555f555f555f555() {
 ; CHECK-NEXT:    ret
   ret i64 u0xf555f555f555f555
 }
+
+define void @test_store_0x1234567812345678(ptr %x) {
+; CHECK-LABEL: test_store_0x1234567812345678:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov x8, #22136 // =0x5678
+; CHECK-NEXT:    movk x8, #4660, lsl #16
+; CHECK-NEXT:    stp w8, w8, [x0, #32]
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/movimm-expand-ldst.ll:102]: I think this should still be a 0 offset, as it is storing to the same place. Can you add a test with a non-zero offset, as in this example:
```
define void @test_store_0x1234567812345678_o(ptr %x) {
; CHECK-LABEL: test_store_0x1234567812345678_o:       
; CHECK:       // %bb.0:                              
; CHECK-NEXT:    mov x8, #22136 // =0x5678            
; CHECK-NEXT:    movk x8, #4660, lsl #16              
; CHECK-NEXT:    stp w8, w8, [x0, #32]                
; CHECK-NEXT:    ret                                  
  %g = getelementptr i64, ptr %x, i64 4               
  store i64 u0x1234567812345678, ptr %g               
  ret void                                            
}  
```

I believe it needs to double the immediate from the str when generating the stp, as the first offset is multiplied by 8, the second by 4. It would be good to test other offsets too, especially the edge-cases close to the boundaries of the allowed values.

Comment (ParkHanbum) [reply llvm/test/CodeGen/AArch64/movimm-expand-ldst.ll:102]: ah! I found it!! how about use 32-bit register like `str` do it? using`STPXi` instead `STPWi` can fix easier.

**Review by davemgreen — COMMENTED** [review]
Thanks - from what I can tell this looks OK now, it's looking good.

I believe that because the two operands are equal this will work for big-endian as well as little endian. Does that sound correct?

**Comment by ParkHanbum** [comment]
honestato be honest, I hadn't considering of that. and I haven't big-endian arm64 machine that can test it in real machine. if you wonder it work or not, I'll find way to check it.

but fortunately in this case, we write just same 4byte value to same address space. so I can say it will work as properly with more confiedence.

**Review by davemgreen — APPROVED** [review]
Yeah I agree about big endian.

I should note that every time we try and implement these backend optimizations at the mir level something ends up going wrong, so please keep an eye out for any problems this might cause.

LGTM

File: llvm/test/CodeGen/AArch64/movimm-expand-ldst.ll [hunk 9]
```diff
@@ -93,3 +93,183 @@ define i64 @testuu0xf555f555f555f555() {
...
+
+define void @test_store_0x1234567812345678_offset_max_over(ptr %x) {
+; CHECK-LABEL: test_store_0x1234567812345678_offset_max_over:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov x8, #22136 // =0x5678
+; CHECK-NEXT:    movk x8, #4660, lsl #16
+; CHECK-NEXT:    orr x8, x8, x8, lsl #32
+; CHECK-NEXT:    stur x8, [x0, #249]
+; CHECK-NEXT:    ret
+  %g = getelementptr i1, ptr %x, i32 249
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply llvm/test/CodeGen/AArch64/movimm-expand-ldst.ll:272]: @davemgreen I really misunderstand what you say before. I was think only i1 type can be increase 1 offset, not now. thanks for let me learn.

**Review by davemgreen — APPROVED** [review]

**Comment by ParkHanbum** [comment]
since https://github.com/ParkHanbum/llvm-project/commit/36231a5b5525b950daf1b7430859061b31a8e01e merged we could not use negative offset for `STR`. therefore two test-cases in .mir were changed.
@davemgreen could you let me know when this PR will be merged?

**Comment by davemgreen** [comment]
Hi - Sorry I hadn't realised this had not been merged. It still LGTM, lets give it a try.

**Comment by vitalybuka** [comment]
Could be this caused by the  patch https://lab.llvm.org/buildbot/#/builders/24/builds/1003 ?

**Comment by chapuni** [comment]
This broke aarch64 stage2. Likely miscompilation.
https://lab.llvm.org/buildbot/#/builders/4/builds/1466

**Comment by vitalybuka** [comment]
> Could be this caused by the patch https://lab.llvm.org/buildbot/#/builders/24/builds/1003 ?

Bisects points to this patch.

**Comment by ParkHanbum** [comment]
@vitalybuka sorry about that. can I ask you a question? is what used in build-test different from test-suite?

**Comment by vitalybuka** [comment]
> ask you a question? is what used in build-test different from test-suite?

It runs with Asan, isn't this report is not enough?

```
FAILED: include/llvm/TargetParser/AArch64TargetParserDef.inc /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm_build_asan/include/llvm/TargetParser/AArch64TargetParserDef.inc 
cd /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm_build_asan && /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm_build_asan/bin/llvm-min-tblgen -gen-arm-target-def -I /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Target/AArch64/ -I /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/include/llvm/TargetParser -I/home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm_build_asan/include -I/home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/include /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Target/AArch64/AArch64.td --write-if-changed -o include/llvm/TargetParser/AArch64TargetParserDef.inc -d include/llvm/TargetParser/AArch64TargetParserDef.inc.d
=================================================================
==263935==ERROR: AddressSanitizer: stack-use-after-scope on address 0xe97677d97b28 at pc 0xbf998ddcbfe8 bp 0xffffe4cb9b30 sp 0xffffe4cb9b28
READ of size 8 at 0xe97677d97b28 thread T0
    #0 0xbf998ddcbfe4 in operator()<unsigned long, unsigned long> /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/libcxx_build_asan/include/c++/v1/__algorithm/comp.h:41:12
    #1 0xbf998ddcbfe4 in min<unsigned long, std::__1::__less<void, void> > /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/libcxx_build_asan/include/c++/v1/__algorithm/min.h:30:10
    #2 0xbf998ddcbfe4 in min<unsigned long> /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/libcxx_build_asan/include/c++/v1/__algorithm/min.h:36:10
    #3 0xbf998ddcbfe4 in llvm::StringRef::find_first_not_of(llvm::StringRef, unsigned long) const /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/StringRef.cpp:265:22
    #4 0xbf998ddc33d0 in ltrim /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/include/llvm/ADT/StringRef.h:783:42
    #5 0xbf998ddc33d0 in trim /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/include/llvm/ADT/StringRef.h:807:14
    #6 0xbf998ddc33d0 in llvm::formatv_object_base::parseReplacementItem(llvm::StringRef) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/FormatVariadic.cpp:60:30
    #7 0xbf998ddc3be4 in llvm::formatv_object_base::splitLiteralAndReplacement(llvm::StringRef) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/FormatVariadic.cpp:135:15
    #8 0xbf998ddc42e0 in llvm::formatv_object_base::parseFormatString(llvm::StringRef) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/FormatVariadic.cpp:151:24
    #9 0xbf998ddf5870 in llvm::formatv_object_base::format(llvm::raw_ostream&) const /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/include/llvm/Support/FormatVariadic.h:86:20
    #10 0xbf998ddf5770 in llvm::raw_ostream::operator<<(llvm::formatv_object_base const&) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/raw_ostream.cpp:351:7
    #11 0xbf998dd2ba30 in EmitARMTargetDef(llvm::RecordKeeper&, llvm::raw_ostream&) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/utils/TableGen/ARMTargetDefEmitter.cpp:270:8
    #12 0xbf998de2bc48 in llvm::TableGenMain(char const*, std::__1::function<bool (llvm::raw_ostream&, llvm::RecordKeeper&)>) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/TableGen/Main.cpp:136:5
    #13 0xbf998dd251c8 in main /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/utils/TableGen/TableGen.cpp:84:10
    #14 0xed7679c284c0  (/lib/aarch64-linux-gnu/libc.so.6+0x284c0) (BuildId: 3119d7bcf68a34e21742cd08095272de601d373f)
    #15 0xed7679c28594 in __libc_start_main (/lib/aarch64-linux-gnu/libc.so.6+0x28594) (BuildId: 3119d7bcf68a34e21742cd08095272de601d373f)
    #16 0xbf998dc3ea2c in _start (/home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm_build_asan/bin/llvm-min-tblgen+0x15ea2c)
Address 0xe97677d97b28 is located in stack of thread T0 at offset 296 in frame
    #0 0xbf998ddc32e8 in llvm::formatv_object_base::parseReplacementItem(llvm::StringRef) /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/FormatVariadic.cpp:59
  This frame has 10 object(s):
    [32, 48) 'ref.tmp.i179'
    [64, 80) 'ref.tmp.i160'
    [96, 112) 'ref.tmp.i129'
    [128, 136) 'ULLVal.i.i'
    [160, 176) 'ref.tmp.i101'
    [192, 200) 'ULLVal.i'
    [224, 240) 'ref.tmp.i82'
    [256, 272) 'ref.tmp.i'
    [288, 304) 'Spec' <== Memory access at offset 296 is inside this variable
    [320, 336) 'RepString' (line 60)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-scope /home/b/sanitizer-aarch64-linux-bootstrap-asan/build/llvm-project/llvm/lib/Support/StringRef.cpp:265:22 in llvm::StringRef::find_first_not_of(llvm::StringRef, unsigned long) const
Shadow bytes around the buggy address:
  0xe97677d97880: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0xe97677d97900: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0xe97677d97980: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0xe97677d97a00: f1 f1 f1 f1 f8 f8 f2 f2 f8 f8 f2 f2 f8 f8 f2 f2
  0xe97677d97a80: f8 f2 f2 f2 f8 f8 f2 f2 f8 f2 f2 f2 f8 f8 f2 f2
=>0xe97677d97b00: 00 00 f2 f2 f8[f8]f2 f2 00 00 f3 f3 00 00 00 00
  0xe97677d97b80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0xe97677d97c00: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0xe97677d97c80: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0xe97677d97d00: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0xe97677d97d80: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==263935==ABORTING
```

**Comment by ParkHanbum** [comment]
@vitalybuka What I mean is that if you can do the same tests, it won't make mistakes. I think there must be something wrong with my testing because when I ran the test-suite I didn't find any issues.

**Comment by vitalybuka** [comment]
test-suite has nothing to do with https://lab.llvm.org/buildbot/#/builders/24/builds/1003, we don't run the test-suit there, we build compiler with compiler.

And this report looks bogus. So asan is broken with your patch.

This explains how to reproduce, if you have aarch64 workstation https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild

TLDR
`stage1 build all` builds regular clang from the HEAD, including your patch
`stage2/asan build` uses that clang from stage1 to build clang instrumented with asan, there is tblg gen stape in this build, and it uses just build clang_with_asan, and it produces false reports.

**Comment by ParkHanbum** [comment]
@vitalybuka thanks. it is helpful. I'll try it first at next time.

**Comment by gribozavr** [comment]
We have identified (but not reduced) another miscompilation. Thank you @vitalybuka for posting the details of this bootstrap failure. I will be reverting the commit now.

**Comment by gribozavr** [comment]
Ah @vitalybuka already reverted it in https://github.com/llvm/llvm-project/commit/7f7f4feaf07dd3bb4b22d0c25d34b6c99c753aa2 - thank you!

**Comment by chapuni** [comment]
FYI this miscompiles `TEST(DWARFFormValue, SignedConstantForms)` in `llvm/unittests/DebugInfo/DWARF/DWARFFormValueTest.cpp`

**Comment by chapuni** [comment]
I've created a minimal testcase.

```
int f(unsigned long *s, const unsigned long *l) {
  *s = 0x1234567812345678UL;
  return *l == 0x1234567812345678UL;
}
```

Before
```
        mov     x8, #22136                      // =0x5678
        movk    x8, #4660, lsl #16
        orr     x8, x8, x8, lsl #32 // Creates symmetric i64 onto x8
        str     x8, [x0]
        ldr     x9, [x1]
        cmp     x9, x8 // Assumes x8 is symmetric i64
        cset    w0, eq
        ret
```

After
```
        mov     x8, #22136                      // =0x5678
        movk    x8, #4660, lsl #16
        stp     w8, w8, [x0]
        ldr     x9, [x1]
        cmp     x9, x8 // Supposes x8 is symmetric i64
        cset    w0, eq
        ret
```

**Comment by ParkHanbum** [comment]
@chapuni wow!! Thank you very much!!!

**Comment by ParkHanbum** [comment]
@davemgreen can I reopen PR again?

**Comment by chapuni** [comment]
@ParkHanbum Please open another request. :)
